### PR TITLE
Feat/mi add mask

### LINKED ISSUE #3575: [Feature]: masking for mutual information loss
### 🚀 Feature Description

The mutual_information submodule of kornia.losses was created in a former PR. Yet, at the time, we did not implement comparison of signals in restriction to the overlap of there respective regions of interest, as the context may lead to define.

The feature would be an enhancement of the original implementation with optional binary masks for both the reference signal and the signal to which it is compared. If no mask would be provided, the behaviour would be the current one.

### 📂 Feature Category

Loss Functions

### 💡 Motivation

 Typically, the treated signals may suffer detectable artefacts in certain regions   and one may wish to exclude these artefacts in the loss computations. the proposed usage of masks would solve this (via roi_mask = 1 - artefacts_mask).

### 💭 Proposed Solution

update __init__ and forward methods for the relevant classes, update joint_histogram too. Update functional variants.

Please assign me, I think I am almost done with the feature.

### 🔄 Alternatives Considered

I considered relying on the MaskedTensor project, but it appears to be inactive for too long.

### 🎯 Use Cases

_No response_

### 📝 Additional Context

_No response_

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by completementgaga**
## Add masking to mutual information losses

**Closes #3575**

---

## 🛠️ Changes Made
- [x] Add masking everywhere

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** : added test for masking, the former tests still pass
- [ ] **Manual Verification:** (Describe the steps you took)
- [ ] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [X] 🟢 **No AI used.**
- [ ] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

### CODE REVIEW

File: kornia/losses/mutual_information.py [hunk 1]
```diff
@@ -159,7 +173,36 @@ def __init__(
...
+
+        Returns:
+            torch.Tensor: normalized mask
+        """
+        if mask is None:
+            mask = torch.tensor(1, dtype=torch.bool)
+        if mask.ndim > 1:
+            raise ValueError("the mask has to be a common mask for all elements of the batch, not a batch of masks")
+        mask = mask.broadcast_to(masked_guy.shape[-1])
+        return mask
```
[the remark this answers is not in the document]
Comment (completementgaga) [reply kornia/losses/mutual_information.py]: Well, it depends if you use a context manager to specify the device or not.

Yet, I fixed this by explicitly converting in  https://github.com/kornia/kornia/pull/3581/commits/02bcf6748934ee5361c95ff4591b9264e920678b

**Comment by completementgaga** [comment]
I got bitten by the non automatic installation of pre-commit. Couldn't pixi run install also entail this installation? 
I fixed this. Yet, pre-commit.ci still does not pass but for a file I did not touch.

**Comment by completementgaga** [comment]
@edgarriba Please, have a look. I think everything is in order (check https://github.com/kornia/kornia/pull/3581/commits/02bcf6748934ee5361c95ff4591b9264e920678b that solves the remaining complaint of qodo).

**Comment by completementgaga** [comment]
@ducha-aiki @edgarriba Anything I can do to move forward?

**Review by shijianjian — DISMISSED** [review]
Probably using reference_mask as the initialization name is better than mask.

**Comment by completementgaga** [comment]
@shijianjian thanks a lot for taking a look. For the mask name, I understand that when you dive into the inner implementation of forward, reference_mask opposes better to other_mask. Yet upon instantiation of MIFromRef there is no ambiguity regarding which image/signal 'mask'  concerns. So I find things clear enough for the end user as they are.

I also made some merging for the test file (Some  permutations of dtype and device kwargs in certain calls occured between versions)

**Review by shijianjian — APPROVED** [review]
